### ISSUE TITLE
Swarm Jobs Proposal

### ISSUE THREAD

**Issue by dperny** [moby/moby#39447]
This issue is a proposal for the implementation of Swarm Jobs. It has already undergone a closed design review, and I am now bringing it to the community for your input. This proposal has been adapted from that format.

This issue is a formal feature proposal for docker/swarmkit#2852 and #23880.

# Problem Statement

Swarm Jobs are a feature allowing the scheduling of different kinds of one-off Tasks, which run to completion and then exit. Jobs are desired to perform tasks like database migrations, or periodic batch operations. Jobs either run on-demand, or are scheduled to occur at specific times (“cron-jobs”). augmenting Replicated and Global services.

## Background

Swarmkit is the Docker Engine’s native container orchestration platform. It provides a way to easily manage distributed services in a cluster. Swarmkit represents workloads as “Services” which spawn “Tasks”. The user inputs the desired state of the Service as a “Service Spec”, which explains what and how a Service should run. Swarm creates a Service object to represent the actual state of the Service, and then spawns Tasks based on the spec.

Currently, Swarm supports Services in two modes: Replicated and Global. Replicated Services run a certain number of Tasks of a given workload somewhere on the cluster. Global Services run one Task on each node in the cluster (subject to constraints). Both of these Service modes are designed for long-running daemon-type workloads, like web servers or databases.

Swarmkit’s model includes the concept of Task States, which track the progression and lifecycle of a given Task of a Service. When a Task exits, it enters one of several terminal states, depending on what caused it to exit. Regardless of the terminal state entered, with existing Service modes, a new Task is created to replace the Task that exited. With some irrelevant states elided, these terminal states include:

* `COMPLETE`: entered when a task exits with no errors (code 0). 
* `SHUTDOWN`: the task is is requested to exit by the orchestrator.
* `FAILED`: the task exits with an error (code non-0).
* `REJECTED`: the task cannot be executed on the given node.

Notably, the `COMPLETE` state is considered more-or-less equivalent to the other terminal states by most of the swarmkit code, and will result in a new Task being created. However, this state becomes important to the implementation of Jobs, as it allows us to differentiate between successful and unsuccessful job runs.

## Goals

* Implement support for one-off on-demand jobs
* Implement support for scheduled, recurring “cron” jobs.
* Allow specifying total number of desired completed jobs, and total number of simultaneously executing Tasks for a job.
* Allow specifying a “global” job, to execute on every node.
* Allow specifying a failure threshold, beyond which job execution will stop being attempted.
* Allowing re-execution of a completed job.

## Non-Goals
* Provide assistance for coordinating or partitioning work between instances of a Job.
* Implement any sort of dependency between jobs (“job A only proceeds if job B first succeeds”)
* Garbage collection of completed jobs.

# Functional Elements

### Jobs

“Jobs” is not a technical term. A Job is a kind of Service, and it follows the rules of all other Swarm Services. The most notable difference between job Services and the existing long-running Services is that the Tasks spawned for Jobs are not rescheduled when they enter the `COMPLETE` state. Jobs have two modes, like long-running services: Replicated and Global.

#### Replicated Jobs

A Replicated Job is any job for which a certain number of successful runs is expected. Replicated jobs can execute in parallel, and will be scheduled to any node meeting the placement constraints of the Service. Replicated Jobs have two main parameters: maximum concurrent tasks, and total completions.

Maximum concurrent tasks indicates the maximum number of Tasks to execute at the same time for the job. When a Replicated job is started, Swarmkit will launch Tasks up to the count of maximum concurrent tasks. Like existing Services, if placement constraints do not allow this many tasks to be run simultaneously, then the maximum number of tasks may not be the number actually running.

Total completions is the total number of Tasks desired to enter the `COMPLETE` state. Jobs will be fulfilled when the count of `COMPLETE` tasks is equal to the total completions.

When a Task completes, another Task can be scheduled to execute. This will continue until the sum of the total of number of running tasks and the total number of completed tasks is equal to value of total completions. This means a job will never spawn more Tasks than necessary to fulfill the desired total completions.

#### Global Jobs

Global Jobs are like Global Services; they run on every node matching placement constraints. Global Jobs are much simpler than Replicated Jobs, as they have no special parameters. They simply run on every node.

### Running and Re-running Jobs

Creating a job will cause it to be immediately executed. When a job is finished, the Service object will remain until it is manually removed. Users may want to re-execute a job with the same settings many times. To simplify this, users can increment the `ForceUpdate` field of the job’s Task spec, which will cause the job to be re-executed as if it was new. Additionally, any other update to the job’s Spec will cause it to re-execute.

If a job has Tasks running when its execution is requested, those tasks will be immediately terminated, and new Tasks will be started. In the case of cron jobs, if execution of the job’s Tasks takes longer than the scheduling interval, then the Tasks for the job may never complete.

### Cron Jobs

Instead of manually executing and re-executing a job, users may want to schedule a job to run periodically. While this could be accomplished with clever use of the `cron` utility, because of the difficulty of doing so in the distributed cluster environment, this functionality is provided natively within Swarm. Both Replicated and Global jobs can be configured with `cron` -like specificity to execute at particular times or intervals. Importantly, however, this does not allow for `at` -like one-off scheduling at a future time.

Importantly, users may try to set up a Cron Job, and then push new image versions, expecting the job to pull the new image each time. Swarmkit pins the image version by using the image digest when a service is created. This means that each execution of a Cron Job will use the exact same image. Even when Swarm is explicitly instructed not to pin images by digest, this leaves selecting the image version up to the engine, which may or may not pull the latest version of the image. Because of these difficulties, this pattern is not supported.

### Handling Failure

To specify how failed Tasks are handled in Jobs, we used the TaskSpec’s existing `RestartPolicy`. The only difference in behavior for Jobs is that the restart condition `RestartOnAny` will have the same effect as `RestartOnFailure`. The whole goal of a Job is to run to completion, and restarting when it does so is contrary to that purpose.

### Differences from Long-Running Services

Notably, Jobs do not support setting an `UpdateConfig`. The `UpdateConfig` is used to specify the parameters of a rolling update. Jobs do not need to be rolled out; updating a job will trigger its re-execution. If a Job is currently running when it is updated, all running tasks will be aborted.

## REST API

The Docker REST API will be expanded to encompass the new Jobs functionality.

### ServiceMode

The ServiceMode struct indicates the mode of a service. Jobs will introduce two new fields, corresponding to the new modes, to this struct.

```go
// ServiceMode represents the mode of a service.
type ServiceMode struct {
	Replicated    *ReplicatedService `json:",omitempty"`
	Global        *GlobalService     `json:",omitempty"`
	ReplicatedJob *ReplicatedJob     `json:",omitempty"`
	GlobalJob     *GlobalJob         `json:",omitempty"`
}
```

### ReplicatedJob

```go
// ReplicatedJob is a type of one-off job which executes many Tasks in parallel,
// until a certain number of Tasks have succeeded.
type ReplicatedJob struct {
	// MaxConcurrent indicates the maximum number of Tasks that should be
	// executing simultaneously at any given time.
	MaxConcurrent uint64 `json:",omitempty"`
	
	// TotalCompletions sets the total number of Tasks desired to run to
	// completion. This is also the absolute maximum number of jobs that
	// will be executed in parallel. That is, if this number is smaller
	// than MaxConcurrent, only this many replicas will be run in parallel.
	TotalCompletions uint64 `json:",omitempty"`
	
	// Schedule, if non-empty, specifies the times and intervals at which
	// to re-run this job.
	Schedule *Schedule `json:",omitempty"`
}
```

### GlobalJob

```go
// GlobalJob is a type of one-off job which executes one Task on every
// node matching the Service's constraints.
type GlobalJob struct {
	// Schedule, if non-empty, specifies the times and intervals at which
	// to re-run this job.
	Schedule *Schedule `json:",omitempty"`
}
```

### Schedule

The decision of whether to represent the Schedule intervals as a simple string type or as a more structured datatype is a difficult one.

If a cron-like string is used, then it becomes the responsibility of the server to parse that string. Changes to that parser would have to be made very carefully, so as not to break existing strings or change their meaning. On the other hand, new concepts, like non-standard cron characters, could be easily added.

If a structured datatype is used, then there is no concern over parsing, and the cron syntax is not baked into the datatype. The downside is that, to be fully expressive, the data may be represented inefficiently.

```go
// Schedule indicates time intervals. It is a structured datatype which
// represents time intervals as a set of values. Each interval field
// (minutes, hours, day of month, months, day of week) is represented by a
// list of values.
//
// If day of month and day of week are both specified, then only days which
// match *both* values are valid. That is, if the 13th day of the month and
// Friday are both required, then the job will only run on Friday the 13th.
//
// An empty list for any field is invalid (though it may be assigned a valid
// meaning in the future). The equivalent to cron's "*" value is a complete
// list of all possible values for that field.
//
// NOTE: Times specified in the Schedule are _always_ in terms of the
// system timezone. If different nodes have different system timezones
// set, this may lead to odd or unexpected behavior.
type Schedule struct {
	// Minute is the minute of the hour. Valid values are 0-59 inclusive.
	Minute []int
	
	// Hour is the hour of the day. Valid values are 0-23 inclusive.
	Hour []int
	
	// DayOfMonth is the day of the month. Valid values are 1-31 inclusive.
	DayOfMonth []int
	
	// Month is the month of the year. Valid values are 1-12 inclusive.
	Month []int
	
	// DayOfWeek is the day of the week. Valid values are 0-6 inclusive.
	// 0 is Sunday and 6 is Saturday. 7 is not a valid value; yoiu only get
	// one Sunday.
	DayOfWeek []int
	
	// Window represents the window of time in which the job can be executed.
	// Jobs will be executed, by default, as close to the specified time as
	// possible. However, in the case of outages, the Swarm may be unavailable
	// to start execution of the job at precisely the desired time. Window
	// informs Swarmkit up to how long past the the desired scheduling time
	// Swarm should attempt to start the job. If empty or 0, Swarm will never
	// attempt to start the job if its desired execution time has been missed.
	Window time.Duration `json:",omitempty"`
}
```

### JobStatus

The Service object will be amended to include a JobStatus field, which contains an object indicating when the job was last executed and, if a cron job, the computed time when the job will be next executed.

```go
// JobStatus indicates the status of the Service, if it is a Job-mode
// service.
type JobStatus struct {
	// JobIteration is the count of how many times this job has been
	// executed, successfully or otherwise.
	JobIteration Version
	
	// LastExecution is the time that the job was last executed. If
	// the JobStatus is present on the Service, LastExecution will
	// always be present.
	LastExecution time.Time
	
	// NextExecution is the time that the job will next be executed,
	// if the job is a cron job. If the job is not a cron job, this
	// field will be empty.
	NextExecution *time.Time `json:",omitempty"`
}
```

# Design Elaboration

The Design Elaboration contains a lower-level overview of specific details of implementation. It is not necessary to review this section in order to under the feature or its behavior, but it may be useful for providing further detail. 

### Task Life Cycle

The life cycle of a Task in Swarmkit involves the Task passing through several states, and being operated on by specific components concerned with specific states. The life-cycle of job Tasks will be largely identical to that of of service-type tasks, with minor differences.

Tasks are created by the Orchestrator. Each service mode has a corresponding Orchestrator to handle its special requirements. Jobs will likewise need a new Orchestrator to handle the ReplicatedJob and GlobalJob service modes.

These Orchestrators will share much in common with the existing replicated and global orchestrators, but will be much simpler because of the lack of need to handle the service update cases.

When a job is processed by the orchestrator, it will have a job iteration value set. This value will increase monotonically with each successive iteration. Each Task spawned as part of a job will have the job iteration value present, in order to distinguish which iteration of the Task the job belongs to. The existing `SpecVersion` field on Tasks will not be repurposed to store job iteration.

### Scheduling Cron Jobs

In order to handle cron jobs, a new component will be needed. Both Orchestrators will share a common component, here referred to as the “cron scheduler”, which will notify them of the need to reschedule a job. During initialization, the cron scheduler will compute the last desired start time for each job. Then, it will determine if the job has been run since that time. If it has not, and the Window duration has not elapsed, then the cron scheduler will immediately signal the orchestrator to launch the job. In either case, the cron scheduler the computes the duration until the next desired run time, and creates a timer to expire at that time. When one of the timers expires, the cron scheduler will signal the orchestrator to begin the specified job.

The orchestrator will rely on and trust the cron scheduler to notify it of when to execute a job. It will not double-check whether execution of a cron job is needed. Correctness of the cron scheduler is highly important, as if the cron scheduler signals the orchestrator to execute the job erroneously, the job will be executed without question.

## Reliability

Reliability of cron jobs is the most difficult problem in this proposal, as it is the only one that has no analogous solution for the existing service modes. The naive solution for cron jobs, to fire when the scheduling interval has occurred, creates a situation where a system outage during a scheduled cron time can cause the job to not be executed. To ameliorate this problem, the “Window” field is included as part of the Schedule type, and allows the user to specify the window in which execution of the job is valid.

At startup, the cron scheduler will compute the last desired execution time of the job, and compare this to the last known execution time of the job. If no execution has occurred since the last desired interval, and the duration specified by Window has not elapsed since that desired execution time, the cron scheduler will immediately notify the Orchestrator to start the job.

“Time” is a difficult concept in distributed systems, specifically when it comes to ordering events. However, as the whole point of cron jobs is to execute at specific times, we must unfortunately contend with it. To simplify as much as possible, the `JobIteration` field acts as a Lamport clock, associating each job execution with the wall-clock time at which it was started. This way, each Task in a cron job is associated not with the clock-time of the cron job execution but with the particular job iteration, and the responsibility for dealing with wall clock times lies solely in the cron scheduler.

## Scalability

Job functionality, being an extension of existing Service functionality, is subject to the same scalability constraints as existing services. There are a few new scale problems to contend with, however:

* Running cron jobs in a tight (~1 minute) interval will cause heavy loading on the cluster, as Tasks are constantly churned. There is no obvious workaround or optimization for this case, and so users should be advised to not run cron jobs with small intervals.
* Cron jobs are scheduled with minute granularity. This means that by default, users will likely choose minute 0 or minute 1 for every job, and heavy users will likely see large cluster loading at the top of the hour, as all cron jobs will be executed simultaneously. In the future, it may be advisable to invest in functionality analogous to Jenkin’s H character, which instructs the scheduler to pick some time during the interval at random, but use the same time every time, which spreads out the start time of workloads automatically.
* Garbage collection of jobs is excluded from this proposal, which means users will need to ensure they remove jobs that are no longer needed.

**Comment by cpuguy83** [comment]
I would argue, let cron be cron and keep it out of the jobs proposal.
A job scheduler can always be added if it's needed, but there are many tools out there for doing just this, there just isn't a very good way to point such a thing at swarm yet.

Other than having an all-in-one solution, is there some context as to why you want to add scheduled jobs?

**Comment by dperny** [comment]
Of all of the disparate groups I talked to about this feature, cron-like scheduling was quite nearly the only single thing that almost every discussion had in common. It's an important part of this proposal.

**Comment by cpuguy83** [comment]
Sure, but it's not currently possible to schedule jobs on swarm so this will be definition be the thing people ask for.
Support for jobs, and a job scheduler are two very different things.

**Comment by dperny** [comment]
To be clear, "job" just means "run this once" "job scheduler" means "something that executes a job at a particular time", right?

**Comment by dperny** [comment]
Outside of the cron scheduler, could we proceed with the implementation of on-demand jobs?

**Comment by dperny** [comment]
To be clear, this underwent a "closed design review" in the sense that the design was looked at and approved by other engineers, but that decision isn't binding to the open source community, obviously. It was more a step I chose to take to ensure a high quality of the design proposal before I brought it here.

**Comment by dperny** [comment]
I suppose a better idea might be when we revisit the server-side stacks work (it's in progress, just slowly and quietly), that we could implement cron-type scheduling as a part of the server-side stacks code, because that component is intended to be for managing higher-level abstractions _over_ services.

**Comment by cpuguy83** [comment]
Yes I think the rest of the proposal seems solid... not that the job scheduler is not, it's just not a requirement to support jobs and can be handled by other things.

**Comment by dperny** [comment]
I can live with that for now, make progress on jobs and worry about cron later.

**Comment by crazy-max** [comment]
Adding a SkipRunning field like I have done on [swarm-cronjob](https://github.com/crazy-max/swarm-cronjob) could be a good idea for jobs that might take time.

**Comment by SvenDowideit** [comment]
I'd agree with @cpuguy83 that adding Jobs without cron would rock - especially if that means its to get there earlier. SkipRunning _is_ also something that we've needed - we use Swarm to kick off research processing pipelines - and if a model run goes over the job submission period, we'd prefer the existing one to keep going, and we'll figure out what went wrong instead of doubling up.

We have used cronicle in the past as our cron / mgmt ui - and people seem to like it (I have a swarm service plugin that kicks off a run-once service, and then monitors for completion)

**Comment by AkihiroSuda** [comment]
`ReplicatedJob` and `GlobalJob` got implemented in https://github.com/moby/moby/pull/40307

**Comment by SvenDowideit** [comment]
@dperny rockin!

**Comment by trajano** [comment]
I would say to split this into two...

1) single job execution.

This would be triggered via CLI etc.

2) scheduled execution

This should be done through a stack configuration file ideally with a CLI fallback.

The reason for splitting is if the feature was specifically the jobs with no scheduling capability, this will be a smaller scope to implement and hopefully have it released sooner.

The scheduling may be done via plugin but provide some OOTB facilities as well so we don't need to install anything extra. (Which is the reason I like swarm in that most of it is already provided without the need to figure out which plugin to use)

The service would have a deploy like this:

```
jobs:
   myJob:
     image: jobImage
     deploy:
       mode: replicated/global
       schedule_config:
          scheduler: rate/delay/cron/[pluginname] (rate means execute at a given rate)
          options: cronstring/rate/delay time etc
```
       schedule:

**Comment by cpuguy83** [comment]
@trajano That's exactly the route taken. Support for (non-scheduled) jobs is on master.
